libraries/chain/include/eosio/chain/unapplied_transaction_queue.hpp

```diff
@@ -161,23 +184,58 @@ class unapplied_transaction_queue {
       auto itr = queue.get<by_trx_id>().find( trx->id() );
       if( itr == queue.get<by_trx_id>().end() ) {
          fc::time_point expiry = trx->packed_trx()->expiration();
-         queue.insert( { trx, expiry, trx_enum_type::persisted } );
+         size_in_bytes += calc_size( trx );
+         queue.push_back( { trx, expiry, trx_enum_type::persisted } );
       } else if( itr->trx_type != trx_enum_type::persisted ) {
          queue.get<by_trx_id>().modify( itr, [](auto& un){
             un.trx_type = trx_enum_type::persisted;
          } );
       }
    }
 
+   void add_incoming( const transaction_metadata_ptr& trx, bool persist_until_expired, next_func_t next ) {
+      auto itr = queue.get<by_trx_id>().find( trx->id() );
+      if( itr == queue.get<by_trx_id>().end() ) {
+         fc::time_point expiry = trx->packed_trx()->expiration();
+         auto size = calc_size( trx );
+         EOS_ASSERT( size_in_bytes + size < max_transaction_queue_size, tx_resource_exhaustion,
+                     "Transaction ${id} exceeded max transaction queue size: ${size}",
+                     ("id", trx->id())("size", size_in_bytes + size) );
+         size_in_bytes += size;
+         queue.push_back( { trx, expiry, persist_until_expired ? trx_enum_type::incoming_persisted : trx_enum_type::incoming, std::move( next ) } );
+      } else if( (persist_until_expired && itr->trx_type != trx_enum_type::incoming_persisted) || (next && !itr->next) ) {
+         queue.get<by_trx_id>().modify( itr, [persist_until_expired, next{std::move(next)}](auto& un) mutable {
+            if( persist_until_expired ) un.trx_type = trx_enum_type::incoming_persisted;
+            if( next ) un.next = std::move( next );
```

Comment:
Yes. This is really a case that will never happen. I included it for completeness. However, I now think the simpler more clear approach (since this case should never be hit anyway) is to just always replace with given args.